sim-administration/uploader/outfile_to_hss_input_converter.go

```diff
@@ -0,0 +1,336 @@
+ * necessary for this to work are:
+ *
+ *  * The SIM card vendor produces output files similar to the example .out file
+ *     found in the same source directory as this program
+ *
+ *  * The HSS accepts input as a CSV file, with header line 'ICCID, IMSI, KI' and subsequent
+ *    lines containing ICCID/IMSI/Ki fields, all separated by commas.
+ *
+ * Needless to say, the outmost care should be taken when handling Ki values and
+ * this program must, as a matter of course, be considered a security risk, as
+ * must all  software that touch SIM values.
+ *
+ * With that caveat in place, the usage of this program typically looks like
+ * this:
+ *
+ *    ./outfile_to_hss_input_converter.go  \
+ *              -input-file sample_out_file_for_testing.out
+ *              -output-file-prefix  ./hss-input-for-
+ *
+ * (followed by cryptographically strong erasure of the .out file,
+ *  encapsulation of the .csv file in strong cryptography etc., none
+ *  of which are handled by this script).
+ */
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type OutputFileRecord struct {
+	Filename          string
+	inputVariables    map[string]string
+	headerDescription map[string]string
+	entries           []SimEntry
+	noOfEntries       int
+	outputFileName    string
+}
+
+const (
+	INITIAL            = "initial"
+	HEADER_DESCRIPTION = "header_description"
+	INPUT_VARIABLES    = "input_variables"
+	OUTPUT_VARIABLES   = "output_variables"
+	UNKNOWN_HEADER     = "unknown"
+)
+
+type SimEntry struct {
+	rawIccid             string
+	iccidWithChecksum    string
+	iccidWithoutChecksum string
+	imsi                 string
+	ki                   string
+	outputFileName       string
+}
+
+type ParserState struct {
+	currentState      string
+	inputVariables    map[string]string
+	headerDescription map[string]string
+	entries           []SimEntry
+}
+
+func ParseLineIntoKeyValueMap(line string, theMap map[string]string) {
+	var splitString = strings.Split(line, ":")
+	if len(splitString) != 2 {
+		log.Fatalf("Unparsable colon separated key/value pair: '%s'\n", line)
+	}
+	key := strings.TrimSpace(splitString[0])
+	value := strings.TrimSpace(splitString[1])
+	theMap[key] = value
+}
+
+func trimSuffix(s string, suffixLen int) string {
+	return s[:len(s)-suffixLen]
+}
+
+func ReadOutputFile(filename string) OutputFileRecord {
+
+	_, err := os.Stat(filename)
+
+	if os.IsNotExist(err) {
+		log.Fatalf("Couldn't find file '%s'\n", filename)
+	}
+	if err != nil {
+		log.Fatalf("Couldn't stat file '%s'\n", filename)
+	}
+
+	file, err := os.Open(filename) // For read access.
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer file.Close()
+
+	state := ParserState{
+		currentState:      INITIAL,
+		inputVariables:    make(map[string]string),
+		headerDescription: make(map[string]string),
+	}
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
```

Comment:
Maybe trim whitespace at start and at end too?

Reply:
Where are they?   Neither Emacs nor I could find any :-)

Reply:
Was thinking of whitespaces in input...

Reply:
Aaah.  ;-) Thank you (silly me)